OR_GREATER preprocessor symbols for TFMs (#164)
dotnet/designs: accepted/2020/or-greater-defines/or-greater-defines.md

            FastBlurUsingHardwareIntrinsics();
        #else
            SlowBlurUsingSimpleMath();
        #endif
    }
}
```

## Requirements

### Goals

* The new symbols don't change semantics of existing code
* The new symbols allow expressing a `>=` relationship with version numbers
* The new symbols apply to both frameworks as well as platforms
* The new symbols are defined for all frameworks that are relevant for package
  authors today, which includes:
    - .NET 5
    - .NET Core
    - .NET Framework
    - .NET Standard

### Non-Goals

* Making existing code more resilient to TFM changes

## Design

The SDK will define `XXX_OR_GREATER` variants for the following TFMs:

* **.NET Framework**
    - Applies to `net1.0`-`net4x` and the `NETFRAMEWORK`/`NET` symbols
    - For example, `net4.8` will define `NETFRAMEWORK`, `NET48`,
      `NET48_OR_GREATER` `NET472_OR_GREATER`, `NET471_OR_GREATER`, etc.
* .NET Core
    - Applies to `netcoreappX.Y` and the `NETCOREAPP` symbol
    - For example, `netcoreapp3.1` will define `NETCOREAPP`, `NETCOREAPP3_1`,
      `NETCOREAPP3_1_OR_GREATER`, `NETCOREAPP3_0_OR_GREATER` etc.
* .NET 5 and later
    - Applies to `netX.Y` and the `NET` symbol

>     - For example, `net6.0` will define `NET`, `NET6_0`, `NET6_0`, `NET5_0` etc.

Reviewer: I'm not sure what happened here? It makes sense for net6.0 to define NET5_0_OR_GREATER, but not NET5_0.
```suggestion
    - For example, `net6.0` will define `NET`, `NET6_0` etc.
```

Author: That was multi-caret editing mishap. Fixed.

    - This will also include the corresponding defines a .NET Core X.Y would
      have gotten. For example, `net6.0` will also define `NETCOREAPP`,

>       `NETCOREAPP6_0`, `NETCOREAPP5_0_OR_GREATER`, `NETCOREAPP3_1_OR_GREATER`

Reviewer: So NETCOREAPP6_0 is intentionally redundant with NET6_0?

Author: Interesting. I assumed the SDK also defines `NETCOREAPP5_0` when you target `net5.0`. Turns out, it only defines `NETCOREAPP`. I'll update the spec to remove `NETCOREAPP6_0` and `NETCOREAPP5_0_OR_GREATER`.

    #elif NET
        Net5OrHigherBehavior();
    #else
        #error Unhandled TFM
    #endif
}
```

Please note that order is important here. You first handle all the legacy TFMs
and then you handle the new behavior. To avoid breaking that code once you add
target .NET 6, you use the versionless TFM for .NET 5 and higher platforms
(`NET`). However, the code still needs to be revised when you add lower versions
of existing TFMs.

With this feature, you'll be able to express it like this, which is a lot more
intuitive:

```C#
void M()
{
    #if NET5_0_OR_GREATER
        Net5OrHigherBehavior();
    #elif NETCOREAPP3_1_OR_GREATER
        LegacyNetCoreBehavior();
    #elif NET461_OR_GREATER
        LegacyNetFxBehavior();
    #else
        #error Unhandled TFM
    #endif
}
```

The idea is to start from newest to oldest.

This also makes the code more resilient for cases where you add new frameworks
with lower versions later. For example, if you start targeting .NET Core 3.0,
the former code would go down the path for .NET 6 while the latter will fail
with "Unhandled TFM", which is a lot more logical for the person having to
maintain the code.

## Scenarios and User Experience

> ### Multi-targeting for devices to provide a cross-platform API

Reviewer 2: @terrajobst it's not clear how this section relates to the rest of the document. Does `WINDOWS7_0` actually mean `WINDOWS7_0_OR_GREATER`❔ If not and there's no plan to add `WINDOWS7_0_OR_GREATER`, suggest removing this part.

    {
        #if NETSTANDARD2_0
            throw new PlatformNotSupportedException();
        #elif WINDOWS7_0
            return CallWindowsApi();
        #elif ANDROID14_0
            return CallAndroidApi();
        #elif IOS11_0
            return CallIOSApi();
        #else
            #error Provide implementation
        #endif
    }
}
```

### When running on a later version of .NET, use a newer API

Zachariah is building a .NET imaging library. On .NET 5, we have added enough
hardware intrinsics so that he can implement the blur functionality in a much
faster way. However, he still wants to support .NET Standard 2.0, so he decides
to use `#if` to multi-target for .NET 5.

Since he checks for the minimum version of the framework that introduced the
API, adding a new version wouldn't change this code, so he simply puts the
fallback logic for the slow version in the `#else` branch, with no call to
`#error`.

```C#
public class Picture
{
    public void Blur(float radius)
    {
        #if NET5_0_OR_HIGHER
            FastBlurUsingHardwareIntrinsics();
        #else
            SlowBlurUsingSimpleMath();
        #endif
    }
}
```

> ## Requirements

Reviewer 2: Do the requirements include out-of-support TFMs❔ If not, suggest adding that as a non-goal.

Author: The requirements section includes the list of frameworks for which we will do the defines, so I think that's called out.

Author: See my response [here](https://github.com/dotnet/designs/pull/164#issuecomment-723326659). Unless there is a significant cost increase, I think it's easier if do this for all TFMs that can be targeted from the .NET 5 SDK.

# OR_GREATER preprocessor symbols for TFMs

**PM** [Immo Landwerth](https://github.com/terrajobst)

Today, projects automatically define `#if` symbols based on the friendly TFM
representation. This includes both versionless- as well as version-specific
symbols. For .NET 6, we plan to add `XXX_OR_GREATER` variants that are
accumulative in nature:

> TFM                 | Existing | New

Reviewer 3: ```suggestion
TFM                 | Existing | Additional
```

The word "New" here is ambiguous between "this is the new defines we generate that replace the old ones" and "these are the new defines in addition to the older ones". Rest of the doc clears that up but think it could help to be more explicit here.

Author: Reworded to `Proposed Additions`

updated each time the project's `TargetFramework`/`TargetFrameworks` property is
changed. For example, let's say you want a new code path that is triggered for
.NET 6.0 and higher. This could either be a new API you want to call or relying
on a new behavior. Today, you need to write code like that but you need to
consider what version TFMs your project is targeting. Let's say you target .NET
Framework 4.6.1, .NET Core 3.1, and .NET 5.

Today, you'd need to write code like this:

```C#
void M()
{
    #if NET461
        LegacyNetFxBehavior();
    #elif NETCOREAPP3_1
        LegacyNetCoreBehavior();
    #elif NET
        Net5OrHigherBehavior();
    #else
        #error Unhandled TFM
    #endif
}
```

Please note that order is important here. You first handle all the legacy TFMs
and then you handle the new behavior. To avoid breaking that code once you add
target .NET 6, you use the versionless TFM for .NET 5 and higher platforms
(`NET`). However, the code still needs to be revised when you add lower versions
of existing TFMs.

With this feature, you'll be able to express it like this, which is a lot more
intuitive:

```C#
void M()
{

>     #if NET5_0_OR_GREATER

Reviewer 3: VS standard formatting is going to put the `#if` in column 0 but here they are at column 4. Do we want this to look more like the VS experience here?

Reviewer 4: Agreed here. We also use that style for the BCL, and I think we should be consistent.

Author: I find your lack of taste disturbing, but I succumb to the larger principle of consistency :-)

>     #if NET5_0_OR_GREATER

Author: I find your lack of taste disturbing, but I succumb to the larger principle of consistency :-)

>         Net5OrHigherBehavior();

Reviewer 4: Should you be using `Net5OrGreaterBehavior` here, and just generally throughout the document? There appear to be a bunch of mentions of 'Higher' still lingering.

Author: Great catch, fixed.

    #elif NETCOREAPP3_1_OR_GREATER
        LegacyNetCoreBehavior();
    #elif NET461_OR_GREATER
        LegacyNetFxBehavior();
    #else
        #error Unhandled TFM
    #endif
}
```

The idea is to start from newest to oldest.

This also makes the code more resilient for cases where you add new frameworks
with lower versions later. For example, if you start targeting .NET Core 3.0,
the former code would go down the path for .NET 6 while the latter will fail
with "Unhandled TFM", which is a lot more logical for the person having to
maintain the code.

## Scenarios and User Experience

### Multi-targeting for devices to provide a cross-platform API

Joletta is building a cross-platform library for geolocation. Since there is no
built-in API for accessing geolocation, she needs to call different APIs for
different platforms, so she uses `#if` to write different code, depending on the
OS she is building for, plus a .NET Standard implementation that just throws
`PlatformNotSupportedException`.

> Since she needs to provide custom implementations per OS, she wants to fail

Reviewer 4: I don't understand what this has to do with `OR_GREATER`. It's great to have platform ifdefs like that, but they seem unrelated to the rest of the document. A better example in this vein would be making use of some platform-specific APIs only available on a certain OS version.

Author: > I don't understand what this has to do with OR_GREATER. It's great to have platform ifdefs like that, but they seem unrelated to the rest of the document.

Right. One pushback we got for promoting the `OR_GREATER` variants is that some think that it's good that existing code fails when new frameworks/platforms are being targeted. The point of this scenario is to outline how this line of thinkin fits in with this design.

> A better example in this vein would be making use of some platform-specific APIs only available on a certain OS version.

That's what the other scenario covers. `When running on a later version of .NET, use a newer API`. Whether that's OS API or a .NET API doesn't matter.

Reviewer 4: Oh I see, so it's about promoting that pattern rather than the specific defines being added here. I'd still consider putting this as the second scenario in the document; it just seems off to me that the first scenario listed doesn't even use the new defines. Though perhaps this is just a 'me problem' and no one else was confused here :smile:

whenever a new TFM is added:

```C#
public static class Gps
{
    public static GpsCoordinate GetCoordinates()
    {
        #if ANDROID
            return CallAndroidApi();
        #elif IOS
            return CallIOSApi();
        #elif WINDOWS
            return CallWindowsApi();
        #elif NETSTANDARD
            throw new PlatformNotSupportedException();
        #else
            #error Provide implementation
        #endif
    }
}
```

### When running on a later version of .NET, use a newer API

> Zachariah is building a .NET imaging library. On .NET 5, we have added enough

Reviewer 5: Nit

```suggestion
Zachariah is building a .NET imaging library. In .NET 5, we have added enough
```

Author: Fixed

hardware intrinsics so that he can implement the blur functionality in a much
faster way. However, he still wants to support .NET Standard 2.0, so he decides
to use `#if` to multi-target for .NET 5.

Since he checks for the minimum version of the framework that introduced the
API, adding a new version wouldn't change this code, so he simply puts the
fallback logic for the slow version in the `#else` branch, with no call to
`#error`.

```C#
public class Picture
{
    public void Blur(float radius)
    {

>         #if NET5_0_OR_HIGHER

Reviewer 4: ```suggestion
        #if NET5_0_OR_GREATER
```

Author: Fixed

>             FastBlurUsingHardwareIntrinsics();

Reviewer 4: I'm not sure I really like the intrinsics example, mostly because this is architecture-dependent as well and so the checks are always more complicated than this.

Author: Yes and no. Hardward intrinsics have a uniform API, regardless of architecture. The expectation is that the code uses runtime checks via `IsSupported`. The problem here is that if you're using .NET Standard 2.0, you don't have the APIs that allow you write that code in the first place, that's what the `#if` is really solving.

Reviewer 4: Ah gotcha, makes sense. Thanks for clarifying!

* The new symbols are defined for all frameworks that are relevant for package
  authors today, which includes:
    - .NET 5
    - .NET Core
    - .NET Framework
    - .NET Standard

### Non-Goals

* Making existing code more resilient to TFM changes

## Design

The SDK will define `XXX_OR_GREATER` variants for the following TFMs:

* **.NET Framework**
    - Applies to `net1.0`-`net4x` and the `NETFRAMEWORK`/`NET` symbols
    - For example, `net4.8` will define `NETFRAMEWORK`, `NET48`,
      `NET48_OR_GREATER` `NET472_OR_GREATER`, `NET471_OR_GREATER`, etc.
* .NET Core
    - Applies to `netcoreappX.Y` and the `NETCOREAPP` symbol
    - For example, `netcoreapp3.1` will define `NETCOREAPP`, `NETCOREAPP3_1`,
      `NETCOREAPP3_1_OR_GREATER`, `NETCOREAPP3_0_OR_GREATER` etc.
* .NET 5 and later
    - Applies to `netX.Y` and the `NET` symbol
    - For example, `net6.0` will define `NET`, `NET6_0`, `NET6_0_OR_GREATER`,
      `NET5_0_OR_GREATER` etc.
    - This will also include the corresponding defines a .NET Core 3.1 successor
      would have gotten. For example, `net5.0` will also define `NETCOREAPP`,
      `NETCOREAPP3_1_OR_GREATER`, `NETCOREAPP3_0_OR_GREATER` etc. But it will

>       *neither* define a `NETCOREAPP5_0` or `NETCOREAPP5_0_OR_GREATER`.

Reviewer 3: Does this go all the way back to `NETCOREAPP1_1_OR_GREATER`?

Author: > Does this go all the way back to NETCOREAPP1_1_OR_GREATER?

Yes, that's what the `etc.` was supposed to indicate. I've replaced that with an explicit range.

>       *neither* define a `NETCOREAPP5_0` or `NETCOREAPP5_0_OR_GREATER`.

Reviewer 5: ```suggestion
      *neither* define a `NETCOREAPP5_0` nor `NETCOREAPP5_0_OR_GREATER`.
```

* .NET 5 and later with operating systems
    - The OS flavors will also gain the `XXX_OR_GREATER` variants
    - For example, `net5.0-windows10.0.19222.0` will also define `WINDOWS`,
      `WINDOWS10_0_19041_0`, `WINDOWS10_0_19041_0_OR_GREATER`, etc.
* Xamarin

>     - This covers the existing Xamarin offerings, the new .NET 6-based Xamarin

Reviewer 5: ```suggestion
    - This covers the existing Xamarin offerings, the new .NET 6-based iOS and Android support
```

Author: Excellent. Fixed.

      is handled by the previous sections.

>     - The existing Xamarin platforms don't have versioned preprocessor symbols

Reviewer 5: I think there is some level of preprocessor symbols defined for Android API levels.  @mhutch @jonathanpeppers, @Redth

Reviewer 6: Here is where we generate preprocessor symbols on Android:

https://github.com/xamarin/xamarin-android/blob/5fc88a1a7cf67f2045e790b30e43dd7a5a28cd4f/src/Xamarin.Android.Build.Tasks/Tasks/GetAndroidDefineConstants.cs#L28-L38

`__ANDROID_16__` will be present for 16 or higher, for example.

We were planning on keeping these for .NET 6 in addition to any new ones, so we don't break existing code.

Author: I'll add this to the sepc

Author: @Jonathanpeppers @mhutch 

I've spec'ed this as follows. Please let me know if you have any concerns:

> * Xamarin
>    - This covers the existing Xamarin offerings, the new .NET 6-based iOS and
>      Android support is handled by the previous sections.
>    - The existing Xamarin Apple platforms don't have versioned preprocessor
>      symbols, but Android does: `__ANDROID_16__`
>    - We don't plan on adding `OR_GREATER` flavors like
>      `__ANDROID_16_OR_GREATER__`. Instead, we're only doing those for .NET 6.

    - Thus, we won't be adding those

### Doc changes

The [documentation table][docs-table] in the [#if (C# reference)][docs] will

> need to change as follows:

Reviewer 5: The updated table feels too verbose to me.  Instead of having an exhaustive table, perhaps the documentation should have a short description of what preprocessor symbols will be defined as well as examples.

Author: I played with that; the benefit of an exhaustive list is that it leaves no room for interpretation. For example, if someone wonders whether there is `NETCOREAPP6_0_OR_HIGHER` then can easily check that.

> # OR_GREATER preprocessor symbols for TFMs

Reviewer 7: Was there a discussion around the name, e.g. I'd have intuitively called this `OR_LATER` instead.

Author: I proposed OR_LATER before and I was told that OR_GREATER is more consistent with other existing prior art.

### Non-Goals

* Making existing code more resilient to TFM changes

## Design

The SDK will define `XXX_OR_GREATER` variants for the following TFMs:

* **.NET Framework**
    - Applies to `net1.0`-`net4x` and the `NETFRAMEWORK`/`NET` symbols
    - For example, `net4.8` will define `NETFRAMEWORK`, `NET48`,
      `NET48_OR_GREATER`, `NET472_OR_GREATER`..`NET20_OR_GREATER`.
* .NET Core
    - Applies to `netcoreappX.Y` and the `NETCOREAPP` symbol
    - For example, `netcoreapp3.1` will define `NETCOREAPP`, `NETCOREAPP3_1`,
      `NETCOREAPP3_1_OR_GREATER`..`NETCOREAPP1_0_OR_GREATER`.
* .NET 5 and later
    - Applies to `netX.Y` and the `NET` symbol
    - For example, `net6.0` will define `NET`, `NET6_0`, `NET6_0_OR_GREATER` and
      `NET5_0_OR_GREATER`.
    - This will also include the corresponding defines a .NET Core 3.1 successor
      would have gotten. For example, `net5.0` will also define `NETCOREAPP`,
      `NETCOREAPP3_1_OR_GREATER`..`NETCOREAPP1_0_OR_GREATER`.
    - This will *neither* define a `NETCOREAPP5_0` nor
      `NETCOREAPP5_0_OR_GREATER`.
* .NET 5 and later with operating systems
    - The OS flavors will also gain the `XXX_OR_GREATER` variants

>     - For example, `net5.0-windows10.0.19222.0` will also define `WINDOWS`,

Reviewer 7: ```suggestion
    - For example, `net5.0-windows10.0.19041.0` will also define `WINDOWS`,
```

so it matches with the preprocessor symbol examples

      `WINDOWS10_0_19041_0`, `WINDOWS10_0_19041_0_OR_GREATER`, etc.
* Xamarin
    - This covers the existing Xamarin offerings, the new .NET 6-based iOS and
      Android support is handled by the previous sections.
    - The existing Xamarin Apple platforms don't have versioned preprocessor
      symbols, but Android does: `__ANDROID_16__`
    - We don't plan on adding `OR_GREATER` flavors like

>       `__ANDROID_16_OR_GREATER__`. Instead, we're only doing those for .NET 6.

Reviewer 7: Make it clearer that there will be new symbols for Android/iOS:

```suggestion
      `__ANDROID_16_OR_GREATER__`. Instead, we're only adding the symbols described in the previous section for .NET 6.
```

This effectively means we'll be adding `ANDROID_21_OR_GREATER` since the TFM will be `net6.0-android21` right?

Author: Syntax-wise, yes. But I'm not sure where we landed for the TFM versions of Android. AFAIK we weren't using API version but product versions, but that's unrelated to this. The defines would match whatever the TFM version is.

Reviewer 7: We did land on using the API versions, I'll need to update the design doc :)

### Doc changes

The [documentation table][docs-table] in the [#if (C# reference)][docs] will
need to change as follows:

> | Target Frameworks | Symbols |
> | ------------------| ------- |
> | .NET Framework    | `NETFRAMEWORK`, `NET48`, `NET472`, `NET471`, `NET47`, `NET462`, `NET461`, `NET46`, `NET452`, `NET451`, `NET45`, `NET40`, `NET35`, `NET20`, `NET48_OR_GREATER`, `NET472_OR_GREATER`, `NET471_OR_GREATER`, `NET47_OR_GREATER`, `NET462_OR_GREATER`, `NET461_OR_GREATER`, `NET46_OR_GREATER`, `NET452_OR_GREATER`, `NET451_OR_GREATER`, `NET45_OR_GREATER`, `NET40_OR_GREATER`, `NET35_OR_GREATER`, `NET20_OR_GREATER` |
> | .NET Standard     | `NETSTANDARD`, `NETSTANDARD2_1`, `NETSTANDARD2_0`, `NETSTANDARD1_6`, `NETSTANDARD1_5`, `NETSTANDARD1_4`, `NETSTANDARD1_3`, `NETSTANDARD1_2`, `NETSTANDARD1_1`, `NETSTANDARD1_0`, `NETSTANDARD2_1_OR_GREATER`, `NETSTANDARD2_0_OR_GREATER`, `NETSTANDARD1_6_OR_GREATER`, `NETSTANDARD1_5_OR_GREATER`, `NETSTANDARD1_4_OR_GREATER`, `NETSTANDARD1_3_OR_GREATER`, `NETSTANDARD1_2_OR_GREATER`, `NETSTANDARD1_1_OR_GREATER`, `NETSTANDARD1_0_OR_GREATER` |

> > | .NET 6 (and .NET Core) | `NET5_0`, `NETCOREAPP`, `NETCOREAPP3_1`, `NETCOREAPP3_0`, `NETCOREAPP2_2`, `NETCOREAPP2_1`, `NETCOREAPP2_0`, `NETCOREAPP1_1`, `NETCOREAPP1_0`, `NET6_0_OR_GREATER`, `NET5_0_OR_GREATER`, `NETCOREAPP3_1_OR_GREATER`, `NETCOREAPP3_0_OR_GREATER`, `NETCOREAPP2_2_OR_GREATER`, `NETCOREAPP2_1_OR_GREATER`, `NETCOREAPP2_0_OR_GREATER`, `NETCOREAPP1_1_OR_GREATER`, `NETCOREAPP1_0_OR_GREATER` |

Reviewer 7: I wonder if we can leave off the `_0` from .NET5 and later, since we don't plan on releasing minor versions?

It'd save a few characters to type: `NET5_OR_GREATER`.

Reviewer 8: That probably goes hand in hand with [a discussion on Twitter](https://twitter.com/jaredpar/status/1324004651948150784) that happened a few days ago. Basically, `NET40_OR_GREATER` would probably conflict with what `NET40` refers to right now.

Reviewer 7: ah, right, I didn't see that .NET Framework symbols use the shortened name so they'd conflict.

> > | .NET 6 (and .NET Core) | `NET5_0`, `NETCOREAPP`, `NETCOREAPP3_1`, `NETCOREAPP3_0`, `NETCOREAPP2_2`, `NETCOREAPP2_1`, `NETCOREAPP2_0`, `NETCOREAPP1_1`, `NETCOREAPP1_0`, `NET6_0_OR_GREATER`, `NET5_0_OR_GREATER`, `NETCOREAPP3_1_OR_GREATER`, `NETCOREAPP3_0_OR_GREATER`, `NETCOREAPP2_2_OR_GREATER`, `NETCOREAPP2_1_OR_GREATER`, `NETCOREAPP2_0_OR_GREATER`, `NETCOREAPP1_1_OR_GREATER`, `NETCOREAPP1_0_OR_GREATER` |

Reviewer 9: ```suggestion
> | .NET 6 (and .NET Core) | `NET6_0`, `NETCOREAPP`, `NETCOREAPP3_1`, `NETCOREAPP3_0`, `NETCOREAPP2_2`, `NETCOREAPP2_1`, `NETCOREAPP2_0`, `NETCOREAPP1_1`, `NETCOREAPP1_0`, `NET6_0_OR_GREATER`, `NET5_0_OR_GREATER`, `NETCOREAPP3_1_OR_GREATER`, `NETCOREAPP3_0_OR_GREATER`, `NETCOREAPP2_2_OR_GREATER`, `NETCOREAPP2_1_OR_GREATER`, `NETCOREAPP2_0_OR_GREATER`, `NETCOREAPP1_1_OR_GREATER`, `NETCOREAPP1_0_OR_GREATER` |
```

Additionally, why does this define all of `NETCOREAPP3_1`, `NETCOREAPP3_0`, `NETCOREAPP3_1_OR_GREATER`, `NETCOREAPP3_0_OR_GREATER`? That seems incompatible with the note immediately below:

> Version-specific symbols are only defined for the version you're targeting.

It's also incompatible with the current behaviour, where targeting `net5.0` _doesn't_ define `NETCOREAPP3_1`.
